pkg/deviceplugins/vgpu_device_manager.go

```diff
@@ -0,0 +1,499 @@
+
+	err = dp.cleanup()
+	if err != nil {
+		return err
+	}
+
+	sock, err := net.Listen("unix", dp.socketPath)
+	if err != nil {
+		return fmt.Errorf("error creating GRPC server socket: %v", err)
+	}
+
+	dp.server = grpc.NewServer([]grpc.ServerOption{}...)
+
+	pluginapi.RegisterDevicePluginServer(dp.server, dp)
+
+	errChan := make(chan error, 1)
+
+	go func() {
+		errChan <- dp.server.Serve(sock)
+	}()
+
+	err = waitForGRPCServer(dp.ctx, dp.socketPath, connectionTimeout)
+	if err != nil {
+		return fmt.Errorf("error starting the GRPC server: %v", err)
+	}
+
+	err = dp.register()
+	if err != nil {
+		return fmt.Errorf("error registering with device plugin manager: %v", err)
+	}
+
+	dp.setInitialized(true)
+	logrus.Infof("Initialized DevicePlugin: %s", dp.resourceName)
+	dp.starter.started = true
+	err = <-errChan
+
+	return err
+}
+
+func (dp *VGPUDevicePlugin) ListAndWatch(_ *pluginapi.Empty, s pluginapi.DevicePlugin_ListAndWatchServer) error {
+
+	errChan := make(chan error, 1)
+	go func() {
+		errChan <- dp.healthCheck()
+	}()
+
+	emptyList := []*pluginapi.Device{}
+	err := s.Send(&pluginapi.ListAndWatchResponse{Devices: dp.devs})
+	if err != nil {
+		return err
+	}
+	done := false
+	for {
+		select {
+		case devHealth := <-dp.health:
+			for i, dev := range dp.devs {
+				if devHealth.DevID == dev.ID {
+					dp.devs[i].Health = devHealth.Health
+				}
+			}
+			if err := s.Send(&pluginapi.ListAndWatchResponse{Devices: dp.devs}); err != nil {
+				return err
+			}
+			logrus.Debugf("Sending ListAndWatchResponse for device with dpi.devs = %v", dp.devs)
+		case <-dp.stop:
+			done = true
+		case <-dp.done:
+			done = true
+		}
+		if done {
+			break
+		}
+	}
+	// Send empty list to increase the chance that the kubelet acts fast on stopped device plugins
+	// There exists no explicit way to deregister devices
+	if err := s.Send(&pluginapi.ListAndWatchResponse{Devices: emptyList}); err != nil {
+		logrus.Errorf("%s device plugin failed to deregister: %s", dp.resourceName, err)
+	}
+	close(dp.deregistered)
+	return <-errChan
+}
+
+func (dp *VGPUDevicePlugin) Allocate(_ context.Context, r *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
+	logrus.Debugf("Allocate request %s", r.String())
+	resourceNameEnvVar := util.ResourceNameToEnvVar(vgpuPrefix, dp.resourceName)
+	allocatedDevices := []string{}
+	resp := new(pluginapi.AllocateResponse)
+	containerResponse := new(pluginapi.ContainerAllocateResponse)
+
+	for _, request := range r.ContainerRequests {
+		deviceSpecs := make([]*pluginapi.DeviceSpec, 0)
+		for _, devID := range request.DevicesIDs {
+			logrus.Debugf("trying to allocate device for %s", devID)
+			devicePath := filepath.Join(v1beta1.MdevRoot, devID)
+			_, err := os.Stat(devicePath)
+			if err != nil {
+				logrus.Errorf("error allocating device %s: %v", devID, err)
+				continue
+			}
+			deviceSpecs = append(deviceSpecs, &pluginapi.DeviceSpec{
+				ContainerPath: vfioDevicePath,
+				HostPath:      vfioDevicePath,
```

Comment:
Can you help explain why these two values are set to `/dev/vfio`? Thanks

Reply:
they are both the same since we run the pcidevice plugin as a privileged container so it has direct access to the /dev tree on the host.